File: scripts/download_worked_examples.py
#!/usr/bin/env python3
"""
Script to download all Sample Messages from IATA NDC Worked Examples.
Each scenario's messages are saved to a separate folder.
"""

import os
import re
import time
import json
import shutil
import zipfile
import requests
from bs4 import BeautifulSoup
from lxml import etree
from urllib.parse import urljoin, unquote
from pathlib import Path

# Base URL for IATA standards wiki
BASE_URL = "https://standards.atlassian.net"
WORKED_EXAMPLES_URL = f"{BASE_URL}/wiki/spaces/EASD/pages/574586985/Worked+Examples"

# Output directories
# Determine project root based on script location (one level up from scripts/)
PROJECT_ROOT = Path(__file__).resolve().parent.parent

# Output directories
DOWNLOADS_DIR = PROJECT_ROOT / "worked_examples_downloads" # Keep local cache
SITE_DATA_DIR = PROJECT_ROOT / "site/src/data"
SITE_PUBLIC_DIR = PROJECT_ROOT / "site/public/worked_examples"
MAPPING_FILE = SITE_DATA_DIR / "worked_examples.json"

# Version mapping
VERSION_MAP = {
    "2134": "21.3.5",
    "2135": "21.3.5",
    "2136": "24.1",
    "243": "24.3",
    "244": "24.4",
    "245": "25.4",
}

# Request headers to mimic a browser
HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.5",
}


def get_page_content(url: str, session: requests.Session) -> str:
    """Fetch page content with retries."""
    for attempt in range(3):
        try:
            response = session.get(url, headers=HEADERS, timeout=30)
            response.raise_for_status()
            return response.text
        except requests.RequestException as e:
            print(f"  Attempt {attempt + 1} failed: {e}")
            if attempt < 2:
                time.sleep(2)
    return ""



def extract_info(xml_path):
    """Extract message name and version from XML file."""
    try:
        tree = etree.parse(str(xml_path))
        root = tree.getroot()
        
        # Message name
        message_name = root.tag
        if '}' in message_name:
            message_name = message_name.split('}')[-1]
            
        # Version from schemaLocation
        version = None
        schema_location = root.get('{http://www.w3.org/2001/XMLSchema-instance}schemaLocation')
        if schema_location:
            match = re.search(r'xsd/(\d+)/', schema_location)
            if match:
                v_id = match.group(1)
                version = VERSION_MAP.get(v_id, v_id) # Use map or raw id
                
        return message_name, version
    except Exception as e:
        print(f"Error parsing {xml_path}: {e}")
        return None, None


def extract_page_id_from_url(url: str) -> str:
    """Extract page ID from Confluence URL."""
    # URL pattern: /wiki/spaces/EASD/pages/600375364/...
    match = re.search(r'/pages/(\d+)/', url)
    if match:
        return match.group(1)
    return ""


def get_all_scenario_links(session: requests.Session) -> list[dict]:
    """Get all scenario page links from the Worked Examples page."""
    print("Fetching main Worked Examples page...")
    content = get_page_content(WORKED_EXAMPLES_URL, session)
    
    if not content:
        print("Failed to fetch main page!")
        return []
    
    soup = BeautifulSoup(content, 'html.parser')
    scenarios = []
    
    # Find all links within the content area
    # Links to scenario pages contain /pages/ in the URL
    for link in soup.find_all('a', href=True):
        href = link['href']
        
        # Filter for links to pages within EASD space
        if '/wiki/spaces/EASD/pages/' in href and 'Worked+Examples' not in href:
            full_url = urljoin(BASE_URL, href)
            title = link.get_text(strip=True)
            
            # Skip category pages (Offer Worked Examples, Order Worked Examples, etc.)
            if title and 'Worked Examples' not in title:
                page_id = extract_page_id_from_url(full_url)
                if page_id and not any(s['page_id'] == page_id for s in scenarios):
                    scenarios.append({
                        'title': title,
                        'url': full_url,
                        'page_id': page_id
                    })
    
    print(f"Found {len(scenarios)} scenario pages")
    return scenarios


def sanitize_folder_name(name: str) -> str:
    """Convert a title to a valid folder name."""
    # Remove or replace invalid characters
    name = re.sub(r'[<>:"/\\|?*]', '', name)
    # Replace multiple spaces with single space
    name = re.sub(r'\s+', ' ', name)
    # Truncate if too long
    return name[:100].strip()


def download_attachments(scenario: dict, session: requests.Session, metadata: dict, start_count: int) -> int:
    """Download all attachments for a scenario page and update metadata."""
    page_id = scenario['page_id']
    title = scenario['title']
    url = scenario['url']
    
    # Create folder for this scenario
    folder_name = sanitize_folder_name(title)
    folder_path = DOWNLOADS_DIR / folder_name
    folder_path.mkdir(parents=True, exist_ok=True)
    
    # Try to download all attachments as a zip
    download_all_url = f"{BASE_URL}/wiki/download/all_attachments?pageId={page_id}"
    
    current_count = start_count
    downloaded_files = 0
    
    try:
        response = session.get(download_all_url, headers=HEADERS, timeout=60, stream=True)
        
        if response.status_code == 200:
            content_type = response.headers.get('Content-Type', '')
            
            # Check if it's a zip file
            if 'zip' in content_type or 'octet-stream' in content_type:
                zip_path = folder_path / "attachments.zip"
                
                with open(zip_path, 'wb') as f:
                    for chunk in response.iter_content(chunk_size=8192):
                        f.write(chunk)
                
                # Extract the zip file
                try:
                    with zipfile.ZipFile(zip_path, 'r') as zip_ref:
                        zip_ref.extractall(folder_path)
                    
                    # Remove the zip file after extraction
                    zip_path.unlink()
                    
                    # Process extracted files
                    for xml_file in folder_path.glob("*.xml"):
                        msg_name, version = extract_info(xml_file)
                        if not msg_name or not version:
                            continue
                            
                        # Prepare destination
                        target_filename = f"{msg_name}_{version}_{current_count}.xml"
                        shutil.copy(xml_file, SITE_PUBLIC_DIR / target_filename)
                        
                        # Update metadata
                        if msg_name not in metadata:
                            metadata[msg_name] = []
                            
                        metadata[msg_name].append({
                            "title": title,
                            "file": xml_file.name,
                            "url": url,
                            "original_version": version,
                            "path": f"/worked_examples/{target_filename}"
                        })
                        current_count += 1
                        downloaded_files += 1

                    return downloaded_files
                except zipfile.BadZipFile:
                    print(f"    Warning: Downloaded file is not a valid zip")
                    zip_path.unlink()
            else:
                # Not a zip, might be a single file or error page
                # Try alternative approach: scrape page for attachment links
                return download_attachments_individually(scenario, session, folder_path, metadata, start_count)
        else:
            print(f"    Download all returned status {response.status_code}")
            return download_attachments_individually(scenario, session, folder_path, metadata, start_count)
            
    except requests.RequestException as e:
        print(f"    Error downloading: {e}")
        return download_attachments_individually(scenario, session, folder_path, metadata, start_count)
    
    return 0


def download_attachments_individually(scenario: dict, session: requests.Session, folder_path: Path, metadata: dict, start_count: int) -> int:
    """Download attachments individually by parsing the page and update metadata."""
    page_url = scenario['url']
    title = scenario['title']
    url = scenario['url']
    
    content = get_page_content(page_url, session)
    if not content:
        return 0
    
    soup = BeautifulSoup(content, 'html.parser')
    downloaded = 0
    current_count = start_count
    
    # Look for attachment links (usually .xml files)
    for link in soup.find_all('a', href=True):
        href = link['href']
        
        # Confluence attachment URLs contain /attachments/ or /download/attachments/
        if '/attachments/' in href or '/download/' in href:
            if href.endswith('.xml') or '.xml' in href:
                full_url = urljoin(BASE_URL, href)
                
                # Extract filename from URL or link text
                filename = link.get_text(strip=True)
                if not filename.endswith('.xml'):
                    # Try to extract from URL
                    filename = unquote(href.split('/')[-1].split('?')[0])
                
                if filename:
                    try:
                        file_response = session.get(full_url, headers=HEADERS, timeout=30)
                        if file_response.status_code == 200:
                            file_path = folder_path / sanitize_folder_name(filename)
                            with open(file_path, 'wb') as f:
                                f.write(file_response.content)
                                
                            # Process file immediately
                            msg_name, version = extract_info(file_path)
                            if msg_name and version:
                                # Prepare destination
                                target_filename = f"{msg_name}_{version}_{current_count}.xml"
                                shutil.copy(file_path, SITE_PUBLIC_DIR / target_filename)
                                
                                # Update metadata
                                if msg_name not in metadata:
                                    metadata[msg_name] = []
                                    
                                metadata[msg_name].append({
                                    "title": title,
                                    "file": file_path.name,
                                    "url": url,
                                    "original_version": version,
                                    "path": f"/worked_examples/{target_filename}"
                                })
                                current_count += 1
                                downloaded += 1
                                
                    except requests.RequestException as e:
                        print(f"      Failed to download {filename}: {e}")
    
    return downloaded


def main():
    """Main function to orchestrate the download process."""
    print("=" * 60)
    print("IATA NDC Worked Examples Sample Messages Downloader")
    print("=" * 60)
    print()
    
    # Create output directories
    DOWNLOADS_DIR.mkdir(exist_ok=True)
    SITE_DATA_DIR.mkdir(parents=True, exist_ok=True)
    
    # Clear and recreate public directory
    if SITE_PUBLIC_DIR.exists():
        shutil.rmtree(SITE_PUBLIC_DIR)
    SITE_PUBLIC_DIR.mkdir(parents=True, exist_ok=True)
    
    # Create a session for persistent connections
    session = requests.Session()
    
    # Get all scenario links
    scenarios = get_all_scenario_links(session)
    
    if not scenarios:
        print("No scenarios found!")
        return
    
    print()
    print("Starting downloads...")
    print("-" * 60)
    
    total_files = 0
    successful_scenarios = 0
    metadata = {}
    
    for i, scenario in enumerate(scenarios, 1):
        print(f"\n[{i}/{len(scenarios)}] {scenario['title']}")
        print(f"  Page ID: {scenario['page_id']}")
        
        files_saved = download_attachments(scenario, session, metadata, total_files)
        
        if files_saved > 0:
            print(f"  ✓ Processed {files_saved} file(s)")
            total_files += files_saved
            successful_scenarios += 1
        else:
            print(f"  ⚠ No new xml files processed")
        
        # Be polite to the server
        time.sleep(1)
        
    # Save metadata
    with open(MAPPING_FILE, 'w') as f:
        json.dump(metadata, f, indent=2)
    
    print()
    print("=" * 60)
    print("Download Summary")
    print("=" * 60)
    print(f"Total scenarios processed: {len(scenarios)}")
    print(f"Scenarios with downloads: {successful_scenarios}")
    print(f"Total files processed: {total_files}")
    print(f"Output directory: {DOWNLOADS_DIR.absolute()}")
    print(f"Public directory: {SITE_PUBLIC_DIR.absolute()}")
    print(f"Metadata file: {MAPPING_FILE.absolute()}")
    print()


if __name__ == "__main__":
    main()

File: tools/src/main/kotlin/com/ndc/tools/DownloadWorkedExamples.kt
package com.ndc.tools

import com.ndc.tools.common.ExampleSourceRecord
import com.ndc.tools.common.ExampleSourcesFile
import com.ndc.tools.common.IataFlowSourceRecord
import com.ndc.tools.common.IataFlowSourcesFile
import com.ndc.tools.common.IataFlowStepSource
import com.ndc.tools.common.NdcConstants
import com.ndc.tools.common.buildExampleId
import com.ndc.tools.common.contentJson
import com.ndc.tools.common.ensureCanonicalXmlPath
import com.ndc.tools.common.extractMessageInfo
import java.io.File
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.util.zip.ZipFile
import kotlinx.serialization.encodeToString
import kotlin.io.path.createTempDirectory
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element

private const val BASE_URL = "https://standards.atlassian.net"
private const val WORKED_EXAMPLES_URL = "$BASE_URL/wiki/spaces/EASD/pages/574586985/Worked+Examples"

private val USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"

private val httpClient: HttpClient = HttpClient.newBuilder()
    .followRedirects(HttpClient.Redirect.NORMAL)
    .connectTimeout(Duration.ofSeconds(30))
    .build()

data class ScenarioLink(val title: String, val url: String, val pageId: String)

private fun fetchPage(url: String): String {
    for (attempt in 1..3) {
        try {
            val request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("User-Agent", USER_AGENT)
                .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
                .timeout(Duration.ofSeconds(30))
                .GET()
                .build()
            val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
            if (response.statusCode() in 200..299) return response.body()
            println("  Attempt $attempt: HTTP ${response.statusCode()}")
        } catch (e: Exception) {
            println("  Attempt $attempt failed: ${e.message}")
        }
        if (attempt < 3) Thread.sleep(2000)
    }
    return ""
}

private fun fetchBytes(url: String): ByteArray? {
    try {
        val request = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .header("User-Agent", USER_AGENT)
            .timeout(Duration.ofSeconds(60))
            .GET()
            .build()
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofByteArray())
        if (response.statusCode() in 200..299) return response.body()
    } catch (e: Exception) {
        println("    Error downloading: ${e.message}")
    }
    return null
}

fun getAllScenarioLinks(): List<ScenarioLink> {
    println("Fetching main Worked Examples page...")
    val content = fetchPage(WORKED_EXAMPLES_URL)
    if (content.isEmpty()) {
        println("Failed to fetch main page!")
        return emptyList()
    }

    val doc = Jsoup.parse(content, BASE_URL)
    val seen = mutableSetOf<String>()
    val scenarios = mutableListOf<ScenarioLink>()

    for (link in doc.select("a[href]")) {
        val href = link.attr("abs:href").ifEmpty { link.attr("href") }
        if ("/wiki/spaces/EASD/pages/" !in href || "Worked+Examples" in href) continue

        val title = link.text().trim()
        if (title.isEmpty() || "Worked Examples" in title) continue

        val pageIdMatch = Regex("""/pages/(\d+)/""").find(href) ?: continue
        val pageId = pageIdMatch.groupValues[1]

        if (pageId !in seen) {
            seen.add(pageId)
            val fullUrl = if (href.startsWith("http")) href else "$BASE_URL$href"
            scenarios.add(ScenarioLink(title, fullUrl, pageId))
        }
    }

    println("Found ${scenarios.size} scenario pages")
    return scenarios
}

// ── Flow metadata scraping ──

private fun extractSectionText(doc: Document, sectionName: String): String? {
    // Find h2 elements matching the section name
    val header = doc.select("h1, h2, h3").firstOrNull { el ->
        el.text().trim().equals(sectionName, ignoreCase = true)
    } ?: return null

    val parts = mutableListOf<String>()
    var sibling = header.nextElementSibling()
    while (sibling != null && sibling.tagName() !in listOf("h1", "h2")) {
        val text = sibling.text().trim()
        if (text.isNotEmpty()) parts.add(text)
        sibling = sibling.nextElementSibling()
    }
    return parts.joinToString("\n").trim().takeIf { it.isNotEmpty() }
}

private fun extractMainFlowSteps(doc: Document): List<IataFlowStepSource> {
    // Find the Main Flow section header
    val header = doc.select("h1, h2, h3").firstOrNull { el ->
        el.text().trim().equals("Main Flow", ignoreCase = true)
    } ?: return emptyList()

    // Look for a table after the header
    var sibling = header.nextElementSibling()
    var table: Element? = null
    while (sibling != null && sibling.tagName() !in listOf("h1", "h2")) {
        if (sibling.tagName() == "table") {
            table = sibling
            break
        }
        // Also check for tables nested inside divs
        val nested = sibling.selectFirst("table")
        if (nested != null) {
            table = nested
            break
        }
        sibling = sibling.nextElementSibling()
    }

    if (table == null) {
        // Fallback: look for any table on the page with Step/Message/Description columns
        table = doc.select("table").firstOrNull { t ->
            val headers = t.select("th, thead td").map { it.text().trim().lowercase() }
            headers.containsAll(listOf("step", "message"))
        }
    }

    if (table == null) return emptyList()

    val steps = mutableListOf<IataFlowStepSource>()
    val rows = table.select("tr")

    for (row in rows) {
        val cells = row.select("td")
        if (cells.size < 2) continue

        val stepNum = cells[0].text().trim()
        val message = cells[1].text().trim()
        val description = if (cells.size >= 3) cells[2].text().trim() else ""

        // Skip header rows or condition-only rows
        if (stepNum.isEmpty() || message.isEmpty()) continue
        if (stepNum.equals("step", ignoreCase = true)) continue
        if (message.equals("condition", ignoreCase = true)) continue

        // Normalize the message name: add IATA_ prefix if not present
        val normalizedMessage = if (message.startsWith("IATA_")) message
            else "IATA_$message"

        steps.add(IataFlowStepSource(
            stepNumber = stepNum,
            message = normalizedMessage,
            description = description,
        ))
    }
    return steps
}

fun scrapeFlowMetadata(scenario: ScenarioLink, html: String): IataFlowSourceRecord? {
    if (html.isEmpty()) return null

    val doc = Jsoup.parse(html, BASE_URL)

    val description = extractSectionText(doc, "Description")
    if (description.isNullOrBlank()) {
        println("    No Description section found, skipping flow")
        return null
    }

    val steps = extractMainFlowSteps(doc)
    val preconditions = extractSectionText(doc, "Preconditions")
    val postconditions = extractSectionText(doc, "Postconditions")

    return IataFlowSourceRecord(
        id = "flow_iata_page_${scenario.pageId}",
        title = scenario.title,
        description = description,
        sourceUrl = scenario.url,
        sourcePageId = scenario.pageId,
        preconditions = preconditions,
        postconditions = postconditions,
        steps = steps,
    )
}

private fun saveExample(
    xmlFile: File,
    scenario: ScenarioLink,
    examplesDir: File,
    records: MutableMap<String, ExampleSourceRecord>,
): Boolean {
    val (msgName, version) = extractMessageInfo(xmlFile)
    if (msgName == null || version == null) return false

    val id = buildExampleId(
        source = "iata",
        message = msgName,
        version = version,
        sourcePageId = scenario.pageId,
        sourceUrl = scenario.url,
        fileName = xmlFile.name,
    )

    val targetFile = examplesDir.resolve("$id.xml")
    xmlFile.copyTo(targetFile, overwrite = true)

    val record = ExampleSourceRecord(
        id = id,
        source = "iata",
        message = msgName,
        version = version,
        title = scenario.title,
        fileName = xmlFile.name,
        xmlPath = ensureCanonicalXmlPath("ndc_content/examples/files/iata/${targetFile.name}"),
        sourceUrl = scenario.url,
        sourcePageId = scenario.pageId,
        isActive = true,
    )

    records[id] = record
    return true
}

/**
 * Result of processing a single scenario: examples downloaded + optional flow metadata.
 * We cache the fetched page HTML so we don't need to re-fetch for flow scraping.
 */
data class ScenarioResult(
    val filesSaved: Int,
    val pageHtml: String?,
)

private fun downloadAttachments(
    scenario: ScenarioLink,
    examplesDir: File,
    records: MutableMap<String, ExampleSourceRecord>,
): ScenarioResult {
    val tempDir = createTempDirectory("ndc-iata-${scenario.pageId}-").toFile()
    var cachedHtml: String? = null
    try {
        val zipUrl = "$BASE_URL/wiki/download/all_attachments?pageId=${scenario.pageId}"
        val zipBytes = fetchBytes(zipUrl)

        if (zipBytes != null && zipBytes.size > 100) {
            val zipPath = tempDir.resolve("attachments.zip")
            zipPath.writeBytes(zipBytes)

            try {
                ZipFile(zipPath).use { zf ->
                    for (entry in zf.entries()) {
                        if (entry.isDirectory) continue
                        val outFile = tempDir.resolve(entry.name)
                        outFile.parentFile.mkdirs()
                        zf.getInputStream(entry).use { input ->
                            outFile.outputStream().use { output -> input.copyTo(output) }
                        }
                    }
                }
                zipPath.delete()
                val fromZip = processDownloadedFiles(tempDir, scenario, examplesDir, records)
                if (fromZip > 0) {
                    // Fetch page for flow scraping (zip didn't require page fetch)
                    cachedHtml = fetchPage(scenario.url)
                    return ScenarioResult(fromZip, cachedHtml)
                }
            } catch (_: Exception) {
                println("    Warning: Not a valid zip, trying individual downloads")
                zipPath.delete()
            }
        }

        val result = downloadAttachmentsIndividuallyWithHtml(scenario, tempDir, examplesDir, records)
        return result
    } finally {
        tempDir.deleteRecursively()
    }
}

private fun downloadAttachmentsIndividuallyWithHtml(
    scenario: ScenarioLink,
    folderPath: File,
    examplesDir: File,
    records: MutableMap<String, ExampleSourceRecord>,
): ScenarioResult {
    val content = fetchPage(scenario.url)
    if (content.isEmpty()) return ScenarioResult(0, null)

    val doc = Jsoup.parse(content, BASE_URL)
    var downloaded = 0

    for (link in doc.select("a[href]")) {
        val href = link.attr("abs:href").ifEmpty { link.attr("href") }
        if ("/attachments/" !in href && "/download/" !in href) continue
        if (!href.contains(".xml")) continue

        val fullUrl = if (href.startsWith("http")) href else "$BASE_URL$href"
        var filename = link.text().trim()
        if (!filename.endsWith(".xml")) {
            filename = java.net.URLDecoder.decode(href.substringAfterLast("/").substringBefore("?"), "UTF-8")
        }

        if (filename.isNotEmpty()) {
            val fileBytes = fetchBytes(fullUrl) ?: continue
            val filePath = folderPath.resolve(NdcConstants.sanitizeFolderName(filename))
            filePath.writeBytes(fileBytes)
            if (saveExample(filePath, scenario, examplesDir, records)) downloaded++
        }
    }
    return ScenarioResult(downloaded, content)
}

private fun processDownloadedFiles(
    folderPath: File,
    scenario: ScenarioLink,
    examplesDir: File,
    records: MutableMap<String, ExampleSourceRecord>,
): Int {
    var count = 0
    for (xmlFile in folderPath.walkTopDown().filter { it.isFile && it.extension.lowercase() == "xml" }) {
        if (saveExample(xmlFile, scenario, examplesDir, records)) count++
    }
    return count
}

fun main(args: Array<String>) {
    val sep = "=".repeat(60)
    println(sep)
    println("IATA NDC Worked Examples & Flows Downloader -> ndc_content")
    println(sep)
    println()

    val projectRoot = NdcConstants.projectRoot()
    val contentExamplesDir = NdcConstants.examplesRoot(projectRoot).resolve("files/iata")
    val sourceFile = NdcConstants.examplesRoot(projectRoot).resolve("sources/iata.generated.json")
    val flowSourcesDir = NdcConstants.flowSourcesRoot(projectRoot)
    val flowSourceFile = flowSourcesDir.resolve("iata.generated.json")

    sourceFile.parentFile.mkdirs()
    flowSourcesDir.mkdirs()
    if (contentExamplesDir.exists()) contentExamplesDir.deleteRecursively()
    contentExamplesDir.mkdirs()

    val scenarios = getAllScenarioLinks()
    if (scenarios.isEmpty()) {
        println("No scenarios found!")
        return
    }

    println()
    println("Starting downloads...")
    println("-".repeat(60))

    var successfulScenarios = 0
    var totalFiles = 0
    val recordsById = linkedMapOf<String, ExampleSourceRecord>()
    val flowRecords = mutableListOf<IataFlowSourceRecord>()

    for ((i, scenario) in scenarios.withIndex()) {
        println("\n[${i + 1}/${scenarios.size}] ${scenario.title}")
        println("  Page ID: ${scenario.pageId}")

        val result = downloadAttachments(scenario, contentExamplesDir, recordsById)
        if (result.filesSaved > 0) {
            println("  Processed ${result.filesSaved} file(s)")
            totalFiles += result.filesSaved
            successfulScenarios++
        } else {
            println("  No xml files processed")
        }

        // Scrape flow metadata from the page
        val html = result.pageHtml ?: fetchPage(scenario.url)
        val flowRecord = scrapeFlowMetadata(scenario, html)
        if (flowRecord != null) {
            println("  Flow: ${flowRecord.steps.size} steps scraped")
            flowRecords.add(flowRecord)
        } else {
            println("  Flow: no flow metadata found")
        }

        Thread.sleep(1000)
    }

    // Write example sources
    val examples = recordsById.values.sortedWith(compareBy<ExampleSourceRecord> { it.message }.thenBy { it.id ?: "" })
    val output = ExampleSourcesFile(version = 1, examples = examples)
    sourceFile.writeText(contentJson.encodeToString(output))

    // Write flow sources
    val flowOutput = IataFlowSourcesFile(version = 1, flows = flowRecords.sortedBy { it.title })
    flowSourceFile.writeText(contentJson.encodeToString(flowOutput))

    println()
    println(sep)
    println("Download Summary")
    println(sep)
    println("Total scenarios processed: ${scenarios.size}")
    println("Scenarios with downloads: $successfulScenarios")
    println("Total files processed: $totalFiles")
    println("Flows scraped: ${flowRecords.size}")
    println("IATA examples dir: ${contentExamplesDir.absolutePath}")
    println("Metadata file: ${sourceFile.absolutePath}")
    println("Flow sources file: ${flowSourceFile.absolutePath}")
    println()
}
